Add export as csv

tatethurston opened the pull request:
Resolves #2183.

tatethurston commented:
[company.csv](https://github.com/twentyhq/twenty/files/14318742/company.17.csv)

tatethurston commented:
@quest-bot loot #2183

FelixMalfait commented:
Great work! Thanks a lot!!!
You can use `nx database:seed:demo twenty-server` to generate a large number of records, you'll see that the progress computation is a bit off.  (goes slowly 1% ... 2%...3%...4%... then jumps from ~10% to 100% in a matter of ms).
Also it would be nice if you could expand json into columns? E.g. LinkedInLabel, LinkedInUrl as 2 columns. Discard __typename.
Besides that from a product standpoint it's perfect. 
The code looks nice and clean but someone else than me will take a deeper look this week!

tatethurston commented:
Thanks @FelixMalfait I fixed the progress progression and updated the columns

FelixMalfait commented:
Thanks @tatethurston, that's great!
I think there is still a small discrepancy that causes the progress bar to be incorrect. You use PAGE_SIZE = 30 while 60 is the default behavior (you should probably pass `PAGE_SIZE` as a `first` param to make sure it remains consistent if that default changes).

Also, I'm sorry because this wasn't in the initial scope, but do you think it would be possible to export only selected items if there are selected items? And if nothing is selected, keep the current behavior. That'd be amazing!

tatethurston commented:
> Thanks @tatethurston, that's great! I think there is still a small discrepancy that causes the progress bar to be incorrect. You use PAGE_SIZE = 30 while 60 is the default behavior (you should probably pass `PAGE_SIZE` as a `first` param to make sure it remains consistent if that default changes).
> 

Hey @FelixMalfait could you tell me more about the progress bar discrepancy you're noticing? Happy to make the page size setting explicit instead of the current hard coding, but it looks like the API returns 30 records at a time when no limit is set. Atleast, `FindManyCompanies` does.

> > Also, I'm sorry because this wasn't in the initial scope, but do you think it would be possible to export only selected items if there are selected items? And if nothing is selected, keep the current behavior. That'd be amazing!

Yeah I'll take a look at this.

FelixMalfait commented:
@tatethurston sorry hadn't seen the edited version of the comment this morning.
Indeed I hadn't thought of it but if you click anywhere outside the action bar it closes it... Instead I think we then want to add an export button here then https://github.com/twentyhq/twenty/blob/main/packages/twenty-front/src/modules/object-record/record-action-bar/hooks/useRecordActionBar.tsx#L85
But this wasn't part of the initial scope so we can create a separate issue. 

The upper-limit for page size is probably constrained by https://supabase.github.io/pg_graphql/configuration/#max-rows on your side. It is 60 on my side because there's a query to change it upon new workspace creation. https://github.com/twentyhq/twenty/blob/ee7c1fbf5cc91e895cf9946da4f61d43542e3103/packages/twenty-server/src/workspace/workspace-manager/workspace-manager.service.ts#L113 Maybe that isn't the case in the workspace created by seeds?

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -0,0 +1,157 @@
+
+import { useFindManyParams } from '../../hooks/useLoadRecordIndexTable';
+
+export const sleep = (ms: number) =>
```
Should we extract this into a global util file ? With the associated test ?

tatethurston replied:
Up to you guys. I generally wait for another consumer before "promoting" it to a public utility.

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -0,0 +1,157 @@
+      title: col.label,
+    };
+
+    const objectColumnData = rows.find((r) => {
```
Could we add more explicit naming here ? 

- `r` => `row`
- `val` => `fieldValue`
- `const isObject = val && typeof val === 'object' && !Array.isArray(val); return isObject;`

tatethurston replied:
We should be able to push all of this into the dependency now that https://github.com/mrodrig/json-2-csv/issues/247 has landed

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -0,0 +1,157 @@
+      title: col.label,
+    };
+
+    const objectColumnData = rows.find((r) => {
```
`objectColumnData` => `fieldsWithSubFields` ?

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -0,0 +1,157 @@
+        (objectColumnData as any)[column.field],
+      )
+        .filter((key) => key !== '__typename')
+        .map((k) => ({
```
Please avoid one letter variables

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -0,0 +1,157 @@
+  recordIndexId: string;
+};
+
+export const useExportTableData = ({
```
That works.

That's not critical right now, but we should handle the case where we have to stop at maximumRequests. 

I created an issue for that : https://github.com/twentyhq/twenty/issues/4161

lucasbordeau commented:
Note sure that we should export relations: 
<img width="899" alt="image" src="https://github.com/twentyhq/twenty/assets/26528466/1f6ed3ff-6d98-40ed-836c-df3c41c97bed">

Maybe we could export a an array of just label identifier for the relations ? 

Here exporting activity targets for example isn't really relevant.

What we could do as a quick win : 
- Don't export array relations
- Export single relation with their concatenated label identifier : id + label name identifier.

What do you think @FelixMalfait  ?

lucasbordeau commented:
Export only work on one page : 

Working example : 

```tsx
  useEffect(() => {
    (async () => {
      const MAXIMUM_REQUESTS = Math.min(maximumRequests, totalCount / pageSize);

      if (!isDownloading || inflight) {
        return;
      }

      const completeDownload = () => {
        setIsDownloading(false);
        setProgress(undefined);
      };

      const fetchNextPage = async () => {
        setInflight(true);
        await fetchMoreRecords();
        setPageCount((state) => state + 1);
        setProgress(percentage(pageCount, MAXIMUM_REQUESTS));
        await sleep(delayMs);
        setInflight(false);
      };

      console.log({
        hasNextPage,
        pageCount,
        maximumRequests,
        totalCount,
        pageSize,
        MAXIMUM_REQUESTS,
      });

      if (!hasNextPage || pageCount >= MAXIMUM_REQUESTS) {
        csvDownloader(filename, { rows: records, columns });
        completeDownload();
      } else {
        await fetchNextPage();
      }
    })();
  }, [
    delayMs,
    fetchMoreRecords,
    filename,
    hasNextPage,
    inflight,
    isDownloading,
    pageCount,
    records,
    totalCount,
    columns,
    maximumRequests,
    pageSize,
  ]);

```

FelixMalfait commented:
@lucasbordeau yes good point let's avoid exporting relations (solution 1)

charlesBochet commented:
@tatethurston Let us know if you need more input on this one :)

tatethurston commented:
> Export useEffect doesn't fetchMore if we have more than 30 records.
> 
> I think we miss the async/await in our useEffect
> 
> Working example :
> 
> ```tsx
>   useEffect(() => {
>     (async () => {
>       const MAXIMUM_REQUESTS = Math.min(maximumRequests, totalCount / pageSize);
> 
>       if (!isDownloading || inflight) {
>         return;
>       }
> 
>       const completeDownload = () => {
>         setIsDownloading(false);
>         setProgress(undefined);
>       };
> 
>       const fetchNextPage = async () => {
>         setInflight(true);
>         await fetchMoreRecords();
>         setPageCount((state) => state + 1);
>         setProgress(percentage(pageCount, MAXIMUM_REQUESTS));
>         await sleep(delayMs);
>         setInflight(false);
>       };
> 
>       console.log({
>         hasNextPage,
>         pageCount,
>         maximumRequests,
>         totalCount,
>         pageSize,
>         MAXIMUM_REQUESTS,
>       });
> 
>       if (!hasNextPage || pageCount >= MAXIMUM_REQUESTS) {
>         csvDownloader(filename, { rows: records, columns });
>         completeDownload();
>       } else {
>         await fetchNextPage();
>       }
>     })();
>   }, [
>     delayMs,
>     fetchMoreRecords,
>     filename,
>     hasNextPage,
>     inflight,
>     isDownloading,
>     pageCount,
>     records,
>     totalCount,
>     columns,
>     maximumRequests,
>     pageSize,
>   ]);
> ```

Can you expand on what you mean by `Export useEffect doesn't fetchMore if we have more than 30 records` ?

tatethurston commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -0,0 +1,157 @@
+      return val && typeof val === 'object' && !Array.isArray(val);
+    });
+
+    if (objectColumnData) {
```
This will be an object or undefined, not an array

tatethurston commented:
Hey @charlesBochet and @lucasbordeau, two questions:

1. Clarity on this https://github.com/twentyhq/twenty/pull/4034#issuecomment-1963037699
2. How can I test the relation case locally? I don't see people as a selectable field for the company table.

Otherwise I believe I've addressed all of your feedback.

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts:
```diff
@@ -107,18 +111,15 @@ export const useExportTableData = ({
-    }
-
-    const completeDownload = () => {
+    const downloadCav = (rows: object[]) => {
```
What does Cav correspond to ?

lucasbordeau commented:
> Can you expand on what you mean by Export useEffect doesn't fetchMore if we have more than 30 records ?

That's fixed, I couldn't export more than 30 rows.

> How can I test the relation case locally? I don't see people as a selectable field for the company table.

You can add a company on a people (the other way around).

lucasbordeau commented:
@tatethurston Could you remove relation fields entirely from export ? We'll address this in another issue.

lucasbordeau commented:
See https://github.com/twentyhq/twenty/issues/4183

tatethurston commented:
> > Can you expand on what you mean by Export useEffect doesn't fetchMore if we have more than 30 records ?
> 
> That's fixed, I couldn't export more than 30 rows.
> 
> > How can I test the relation case locally? I don't see people as a selectable field for the company table.
> 
> You can add a company on a people (the other way around).

Ah from your screenshot it wasn't clear to me that you were doing a csv export of `people` not of `companies`

tatethurston commented:
Alright @lucasbordeau @charlesBochet @FelixMalfait I believe we're squared away here. Take a look and LMK.

FelixMalfait reviewed (approved):
Perfect! Thanks a lot @tatethurston
